core/src/main/java/com/linecorp/armeria/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.linecorp.armeria.internal.common.RequestContextUtil.newIllegalContextPushingException;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.concurrent.FastThreadLocal;
+
+/**
+ * A {@link RequestContextStorage} which keeps track of {@link RequestContext}s, reporting pushed thread
+ * information if a {@link RequestContext} is leaked.
+ */
+@UnstableApi
+public final class LeakTracingRequestContextStorage implements RequestContextStorage {
+
+    private final RequestContextStorage delegate;
+    private final FastThreadLocal<PendingRequestContextStackTrace> pendingRequestCtx;
+    private final Sampler<Object> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate) {
+        this(delegate, Flags.verboseExceptionSampler());
+    }
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                            Sampler<?> sampler) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        pendingRequestCtx = new FastThreadLocal<>();
+        this.sampler = (Sampler<Object>) requireNonNull(sampler, "sampler");
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
+
+        final RequestContext prevContext = delegate.currentOrNull();
+        if (prevContext != null) {
```

Comment:
cc. other maintainers @ikhoon @minwoox @trustin

Reply:
> At the same time, there are some corner cases where this filter isn't comprehensive

I see. Thanks for pointing this out 🙏.

> Re: Related contexts are not recorded - How about storing whether the current context is sampled? For example, we could always record when `prevContext` or its root is sampled, just like how distributed tracing is propagated to its subspans

Hi @trustin, Could help elaborate on this option? thanks.

Another option that I can think of is to store a stack of stack traces. Actually, @minwoox mentioned in the issue that stack is needed because context can be pushed multiple times before getting popped.

Reply:
> Actually, @minwoox mentioned in the issue that stack is needed because context can be pushed multiple times before getting popped.

Yeah, I think we need a stack to store stack traces. A request context could be pushed multiple times before it's popped, so the stack is the right data structure for this case.
We might be able to do it something like:
```java
private final FastThreadLocal<Deque<ContextAndStackTrace>> threadLocalStack = new FastThreadLocal<>();

@Nullable
@Override
public <T extends RequestContext> T push(RequestContext toPush) {
    ...

    final RequestContext prevContext = delegate.currentOrNull();
    if (prevContext != null) {
        if (...) {
            // valid
        } else {
            throw newIllegalContextPushingException(toPush, prevContext, threadLocalStack.get());
        }
    }
    
    ...
    threadLocalStack.get().push(new ContextAndStackTrace(toPush, currentThread().getStackTrace()));
    ...
}

@Override
public void pop(RequestContext current, @Nullable RequestContext toRestore) {
    final ContextAndStackTrace last = threadLocalStack.get().peekLast();
    if (last == null || last.ctx != current) {
        // exception!
    }
    threadLocalStack.get().pop();
    delegate.pop(current, toRestore);
}

private static class ContextAndStackTrace {

    private final RequestContext ctx;
    private final StackTraceElement[] stackTrace;

    ContextAndStackTrace(RequestContext ctx, StackTraceElement[] stackTrace) {
        this.ctx = ctx;
        this.stackTrace = stackTrace;
    }
}

```

> Hi @trustin, Could help elaborate on this option? thanks.

Let's suppose this situation:
```
a.push()
  b.push() // b.root() == a
    c.push() // c.root() == a
a.pop() // exception! because b and c are not popped.
```
Because b and c are not popped we need to provide the stack traces of those contexts. However, if b is sampled and c isn't sampled we cannot provide them.
In order to solve this situation, I think we need to make a decision to sample or not only when the root of the context is pushed so that we can have all stack traces of the context flow. 😉